### ISSUE TITLE
Do not default health endpoints

### ISSUE THREAD

**Issue by matejvasek** [knative/func#1839]
This is opposite of https://github.com/knative/func/issues/1281.

The defaulting means user cannot unset health endpoints in any way.

The endpoints should be set explicitly in templates.
If the endpoints are not explicitly set in `func.yaml` then knative service should not have health endpoint set.

**Comment by matejvasek** [comment]
cc @mvinkler

**Comment by lance** [comment]
> If the endpoints are not explicitly set in func.yaml then knative service should not have health endpoint set.

**Comment by matejvasek** [comment]
@lance  What if user doesn't have endpoints, for instance when trying to migrate existing project to use func?
IMO user should have ability to not have endpoints.

**Comment by matejvasek** [comment]
>Are you also saying that the frameworks, e.g. faas-js-runtime disable these endpoints if the value is not found in func.yaml?

Nope. The application still may have health endpoints, but knative serving would not be checking them if not set in `func.yaml`.

**Comment by matejvasek** [comment]
This has been discovered by @mvinkler . He tried to port some existing Quarkus code to use func but it did not work because of missing endpoints.

**Comment by matejvasek** [comment]
Defaulting:
  \+ Three fewer lines in `func.yaml`.
  \- There is absolutely no way to disable health probes should user want it.
No Defaulting:
  \+ User can disable health probes should they need it.
  \- There are there more lines in `func.yaml`.

**Comment by mvinkler** [comment]
Thanks Matej for filing this issue. 

Reasoning is the following: 
If user decides to use his existing Quarkus app and includes func.yaml in the project (with default values), health checks will be the point of failure for the app deployment. When the ksvc is created on the cluster, functions project "magically" adds the health checks for readiness and liveness and it doesn't care, if there is smallrye extension with proper paths in application.properties added in the project.

Compare with the default behavior of `kn service create`: it adds the following "minimal" health check:
```
readinessProbe:
  successThreshold: 1
  tcpSocket:
    port: 0
```
User has no way of knowing that there is some default value applied on Functions side and he actually cannot in any way use the mentioned minimal health check.

Addition to Matej's summary:
Defaulting:
- Not consistent with `kn service create`
- User has no way of knowing what happens, where the health checks come from, how to change them.
- No health checks description in Functions documentation.

No Defaulting:
+ The explicitly set health checks are self-documentary.
+ No magic creating health checks on application deploy.

**Comment by mvinkler** [comment]
I hit this using Quarkus, I have no idea if other runtimes also have some extensions similar to Smallrye and the same behavior applies to them.

**Comment by lance** [comment]
@mvinkler @matejvasek thanks for the explanation - that all makes sense

**Comment by lkingland** [comment]
Thank you for considering the situation, opening the topic, and providing a thread of pros/cons!

tl;dr: this is actually a feature request for the Quarkus Functions Runtime, but let's keep the issue open because the discussion is good.   Let me explain:

> [we] tried to port some existing Quarkus code to use func but it did not work because of missing endpoints.

Implementing these endpoints is the responsibility of the Functions middleware (runtime) and an area of active development (see the [Scaffolding Architectural Discussion](https://github.com/knative/func/discussions/1589) and [Go MVC Epic](https://github.com/knative/func/issues/1703)).   If the Function author has implemented health endpoints in their Function, those should be used (and how to do so clearly documented in multiple places; see below).  If they did not implement the endpoints, the middlware will answer those health and readiness checks, and these endpoints should be easily configurable.   This is clearly superior to a hard-coded TCP probe in the func.yaml on project creation for several reasons, and a point in favor of Functions as a "value add" over standard services.

However, this thread does indeed bring up two things worth covering in this week's call:  1) we need better func.yaml documentation and 2) we need to more strictly interpret func.yaml to help avoid confusion confusion.

I will be sure to bring this up in this week's meeting.

Here, I'll address the points made in this thread so far; just to be sure we have a productive conversation.

> @lance What if user doesn't have endpoints, for instance when trying to migrate existing project to use func?

Then the middleware will implement it, to the best of its ability, entirely transparently to the Function developer.

> Defaulting:
+ Three fewer lines in func.yaml.

This is a purposeful mischaracterization of the motivation, intended to cast it as of trivial benefit, and thus insinuating it's equally trivial to change the other way to achieve the subsequent benefits and avoid the cons.  Perhaps it's worth reiterating the actual purpose of the in-code defaults:

By hard-coding the values into the func.yaml when a Function is created, we are inhibiting the user's ability to migrate in the future, and our ability to respond quickly to everything from platform improvements to CVEs.  Updating func (library or binary) updates defaults, allowing the infrastructure supporting a user's Function to grow, evolve, and patch itself intelligently (see: our burgeoning conversation around how to implement a Function's operator).  Updating func binary can not touch the user's func.yaml without possibly destroying user changes, because there is no way to know if a value written to func.yaml simply the default written at Function craetion time, or a value intentionally added by the author. 

Let's consider the next point as an example:

> Defaulting:
>  - There is absolutely no way to disable health probes should user want it.

This is a simple misunderstanding about the purpose of func.yaml.  It should be strictly interpreted as overrides to default middleware behavior.  With that expectation, setting the values to empty should disable the endpoints entirely, though this makes no sense to do, presuming the middleware works correctly.

> − Not consistent with kn service create
See above, an intelligently implemented default endpoint is better than a TCP probe, and a value-add of Functions.

> − User has no way of knowing what happens, where the health checks come from, how to change them.
This is a documentation issue, easily solved with a comment block in func.yaml showing the syntax and current default values.   This is idiomatic of many system config files.

> − No health checks description in Functions documentation.
Duplicate of issue above:  need docs.

> + The explicitly set health checks are self-explanatory.

Again, a documentation issue, not a point either for or against setting defaults in code vs in the config file. 

> + No magic creating health checks on application deploy.

Again, a documentation issue.  Countless things happen outside the purview of a service author.  What causes angst (and levying the prejoritive term "magic") is when complexity that the author _needs_ to know is hidden.  I empathize, and empthatically agree that systems should not include (thusly defined) magic.    The root of the problem here, I believe, is taht the Quarkus Function in question _needed_ to know.  These endpoints should be entirely optional (ignoring them should not cause the Quarkus function to fail), and it should be well documented, in multiple places, how an author can implement their own.

So summarize the conclusion:  it is absolutely essential that we 1) have better documentation of func.yaml  2) implement it trictly to enable complete control by the Function author and 3) write only user-provided overrides to func.yaml.  This latter point is essential to our future flexibility.

Thank you very much for the feedback

**Comment by lkingland** [comment]
Thanks @mvinkler.  In our most recent meeting we discussed using a template for the func.yaml rather than a "blind" marshalling, which will have all possible settings enumerated (but commented out) with their static defaults shown in the comment.  If the user has overridden any of the values, the comment will be omitted in favor of the user-supplied setting.

This method is not without its annoyances (for example, any manual edits to the comments will not be retained), but this should allow one to easily scan through a func.yaml to see what options are available, and uncomment the lines to provide customized overrides.

This will be a good deal of manual work, but its documentation we need to have!

I also agree we should have an easy way to display the effective values on deploy.  Either via an echo during `kn func deploy`, a much -expanded `kn func status`, or both.

Let's leave this issue open until it is superceded by that task/issue directly 👍🏻

**Comment by acelinkio** [comment]
My thoughts on this issue are that having an opinionated implementation is perfectly fine as long as its documented.  Pushing people towards best practices is much better than letting people have absolute freedom.  (Rust borrow checker vs C entirely unrestrained)

I am not sure if there is an open standard between the different function providers.  OpenFaaS, AWS Lambda, etc.
